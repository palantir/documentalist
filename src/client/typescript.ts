/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { IJsDocTags } from "ts-quick-docs";
import { IBlock } from "./index";

export interface ITsDocEntry {
    documentation: IBlock;
    fileName?: string;
    name: string;
    tags?: IJsDocTags;
    type: string;
}

export interface ITsPropertyEntry extends ITsDocEntry {
    optional?: boolean;
}

export interface ITsInterfaceEntry extends ITsDocEntry {
    extends?: string[];
    properties: ITsPropertyEntry[];
}

export interface ITypescriptPluginData {
    ts: {
        [name: string]: ITsInterfaceEntry;
    };
}
