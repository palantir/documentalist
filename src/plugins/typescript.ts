/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import tsdoc, { IDocumentationOptions, IJsDocTags } from "ts-quick-docs";
import { IBlock, ICompiler, IFile, IPlugin } from "./plugin";

export interface ITsDocEntry {
    documentation: IBlock;
    fileName: string;
    name: string;
    tags: IJsDocTags;
    type: string;
}

export interface ITsPropertyEntry extends ITsDocEntry {
    optional?: boolean;
}

export interface ITsInterfaceEntry extends ITsDocEntry {
    extends?: string[];
    properties: ITsPropertyEntry[];
}

export interface ITypescriptPluginData {
    ts: {
        [name: string]: ITsInterfaceEntry;
    };
}

export class TypescriptPlugin implements IPlugin<ITypescriptPluginData> {
    public constructor(
        /**
         * Options to `ts-quick-docs`, mostly for customizing which symbols appear in the output.
         */
        private options: IDocumentationOptions = {},

        /**
         * Compiler options for Typescript program used to "read" your typings.
         * (This is distinct from whatever options you need to build your typings.)
         * If omitted, the default compiler options are used.
         */
        // HACK: using any to avoid duplicate typings issue with ts.CompilerOptions
        private compilerOptions: any = {},
    ) {}

    public compile(files: IFile[], { renderBlock, objectify }: ICompiler) {
        const entries = tsdoc.fromFiles(files.map((f) => f.path), this.compilerOptions, this.options)
            .map<ITsInterfaceEntry>((entry) => ({
                ...entry,
                documentation: renderBlock(entry.documentation),
                properties: entry.properties!.map<ITsPropertyEntry>((prop) => ({
                    ...prop,
                    documentation: renderBlock(prop.documentation),
                })),
            }));
        const ts = objectify(entries, (e) => e.name);
        return { ts };
    }
}
