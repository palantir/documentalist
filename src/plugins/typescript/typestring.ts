/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { SignatureReflection } from "typedoc";
import {
    IntersectionType,
    ReferenceType,
    ReflectionKind,
    ReflectionType,
    Type,
    UnionType,
} from "typedoc/dist/lib/models";

export function resolveTypeString(type: Type): string {
    if (type instanceof ReflectionType) {
        // reflection types include generics and object index signatures
        return type.declaration
            .getAllSignatures()
            .map(resolveSignature)
            .join(" | ");
    } else if (type instanceof UnionType) {
        return type.types.map(resolveTypeString).join(" | ");
    } else if (type instanceof IntersectionType) {
        return type.types.map(resolveTypeString).join(" & ");
    } else if (type instanceof ReferenceType) {
        const name = type.name === "__type" ? "{}" : type.name;
        const typeArgs = type.typeArguments == null ? "" : `<${type.typeArguments.map(resolveTypeString).join(", ")}>`;
        return name + typeArgs;
    } else {
        return type.toString();
    }
}

export function resolveSignature(sig: SignatureReflection): string {
    const { parameters = [] } = sig;
    const paramList = parameters.map(param =>
        // [...]name[?]: type
        [
            param.flags.isRest ? "..." : "",
            param.name,
            param.flags.isOptional || param.defaultValue ? "?" : "",
            ": ",
            resolveTypeString(param.type),
        ].join(""),
    );
    const returnType = resolveTypeString(sig.type);
    switch (sig.kind) {
        case ReflectionKind.CallSignature:
        case ReflectionKind.ConstructorSignature:
            return `(${paramList.join(", ")}) => ${returnType}`;
        case ReflectionKind.IndexSignature:
            return `{ [${paramList}]: ${returnType} }`;
        default:
            return sig.toString();
    }
}
