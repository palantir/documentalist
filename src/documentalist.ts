/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import * as fs from "fs";
import * as glob from "glob";
import * as path from "path";
import { IFile, IPlugin } from "./client";
import { Compiler, ICompilerOptions } from "./compiler";

/**
 * Plugins are stored with the regex used to match against file paths.
 */
export interface IPluginEntry<T> {
    pattern: RegExp;
    plugin: IPlugin<T>;
}

export class Documentalist<T> {
    /**
     * Construct a new `Documentalist` instance with the provided options.
     * This method lends itself particularly well to the chaining syntax:
     * `Documentalist.create(options).use(...)`.
     */
    public static create(options?: ICompilerOptions): Documentalist<{}> {
        return new Documentalist(options, []);
    }

    constructor(private options: ICompilerOptions = {}, private plugins: Array<IPluginEntry<T>> = []) {}

    /**
     * Adds the plugin to Documentalist. Returns a new instance of Documentalist
     * with a template type that includes the data from the plugin. This way the
     * `documentFiles` and `documentGlobs` methods will return an object that is
     * already typed to include each plugin's output.
     *
     * The plugin is applied to all files whose absolute path matches the
     * supplied pattern.
     *
     * @param pattern - A regexp pattern or a file extension string like "js"
     * @param plugin - The plugin implementation
     * @returns A new instance of `Documentalist` with an extended type
     */
    public use<P>(pattern: RegExp | string, plugin: IPlugin<P>): Documentalist<T & P> {
        if (typeof pattern === "string") {
            pattern = new RegExp(`${pattern}$`);
        }

        const newPlugins = [...this.plugins, { pattern, plugin }] as Array<IPluginEntry<T & P>>;
        return new Documentalist(this.options, newPlugins);
    }

    /**
     * Returns a new instance of Documentalist with no plugins.
     */
    public clearPlugins(): Documentalist<{}> {
        return new Documentalist<{}>(this.options, []);
    }

    /**
     * Finds all files matching the provided variadic glob expressions and then
     * runs `documentFiles` with them, emitting all the documentation data.
     */
    public async documentGlobs(...filesGlobs: string[]) {
        const files = this.expandGlobs(filesGlobs);
        return this.documentFiles(files);
    }

    /**
     * Iterates over all plugins, passing all matching files to each in turn.
     * The output of each plugin is merged to produce the resulting
     * documentation object.
     *
     * The return type `T` is a union of each plugin's data type.
     */
    public async documentFiles(files: IFile[]) {
        const compiler = new Compiler(this.options);
        // need an empty object of correct type that we can merge into
        // tslint:disable-next-line:no-object-literal-type-assertion
        const documentation = {} as T;
        for (const { pattern, plugin } of this.plugins) {
            const pluginFiles = files.filter(f => pattern.test(f.path));
            const pluginDocumentation = await plugin.compile(pluginFiles, compiler);
            this.mergeInto(documentation, pluginDocumentation);
        }
        return documentation;
    }

    /**
     * Expands an array of globs and flatten to a single array of files.
     */
    private expandGlobs(filesGlobs: string[]): IFile[] {
        return filesGlobs
            .map(filesGlob => glob.sync(filesGlob))
            .reduce((a, b) => a.concat(b))
            .map(fileName => {
                const absolutePath = path.resolve(fileName);
                return {
                    path: absolutePath,
                    read: () => fs.readFileSync(absolutePath, "utf8"),
                };
            });
    }

    /**
     * Shallow-merges keys form source into destination object (modifying it in the process).
     */
    private mergeInto(destination: T, source: T) {
        for (const key in source) {
            if (source.hasOwnProperty(key)) {
                if (destination.hasOwnProperty(key)) {
                    console.warn(`WARNING: Duplicate plugin key "${key}". Your plugins are overwriting each other.`);
                }
                destination[key] = source[key];
            }
        }
        return destination;
    }
}
