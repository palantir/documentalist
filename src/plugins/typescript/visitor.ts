/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { relative } from "path";
import {
    DeclarationReflection,
    ParameterReflection,
    ProjectReflection,
    Reflection,
    ReflectionKind,
    SignatureReflection,
} from "typedoc";
import { Comment } from "typedoc/dist/lib/models/comments/comment";
import { DefaultValueContainer } from "typedoc/dist/lib/models/reflections/abstract";
import { ICompiler } from "../../client/compiler";
import {
    ITsClass,
    ITsConstructor,
    ITsDocBase,
    ITsFlags,
    ITsInterface,
    ITsMethod,
    ITsMethodParameter,
    ITsMethodSignature,
    ITsProperty,
    Kind,
} from "../../client/typescript";
import { ITypescriptPluginOptions } from "./index";
import { resolveSignature, resolveTypeString } from "./typestring";

export class Visitor {
    public constructor(private compiler: ICompiler, private options: ITypescriptPluginOptions) {}

    public visitProject(project: ProjectReflection) {
        // get top-level members of typedoc project
        const interfaces = this.visitChildren(
            project.getReflectionsByKind(ReflectionKind.Interface),
            this.visitInterface,
        );
        const classes = this.visitChildren(project.getReflectionsByKind(ReflectionKind.Class), this.visitClass);

        // remove members excluded by path option
        const { excludePaths = [] } = this.options;
        return [...interfaces, ...classes].filter(ref => isNotExcluded(excludePaths, ref.fileName));
    }

    private makeDocEntry<K extends Kind>(def: Reflection, kind: K) {
        return {
            documentation: this.renderComment(def.comment),
            fileName: getFileName(def),
            flags: getFlags(def),
            kind,
            name: def.name,
        };
    }

    private visitClass = (def: DeclarationReflection): ITsClass => ({
        ...this.visitInterface(def),
        constructorType: this.visitChildren(
            def.getChildrenByKind(ReflectionKind.Constructor),
            this.visitConstructor,
        )[0],
        kind: Kind.Class,
    });

    private visitInterface = (def: DeclarationReflection): ITsInterface => ({
        ...this.makeDocEntry(def, Kind.Interface),
        extends: def.extendedTypes && def.extendedTypes.map(resolveTypeString),
        implements: def.implementedTypes && def.implementedTypes.map(resolveTypeString),
        methods: this.visitChildren(def.getChildrenByKind(ReflectionKind.Method), this.visitMethod, sortStaticFirst),
        properties: this.visitChildren(
            def.getChildrenByKind(ReflectionKind.Property),
            this.visitProperty,
            sortStaticFirst,
        ),
    });

    private visitConstructor = (def: DeclarationReflection): ITsConstructor => ({
        ...this.visitMethod(def),
        kind: Kind.Constructor,
    });

    private visitProperty = (def: DeclarationReflection): ITsProperty => ({
        ...this.makeDocEntry(def, Kind.Property),
        defaultValue: getDefaultValue(def),
        inheritedFrom: def.inheritedFrom && resolveTypeString(def.inheritedFrom),
        type: resolveTypeString(def.type),
    });

    private visitMethod = (def: DeclarationReflection): ITsMethod => ({
        ...this.makeDocEntry(def, Kind.Method),
        inheritedFrom: def.inheritedFrom && resolveTypeString(def.inheritedFrom),
        signatures: def.signatures.map(sig => this.visitSignature(sig)),
    });

    private visitSignature = (sig: SignatureReflection): ITsMethodSignature => ({
        ...this.makeDocEntry(sig, Kind.Signature),
        flags: undefined,
        parameters: (sig.parameters || []).map(param => this.visitParameter(param)),
        returnType: resolveTypeString(sig.type),
        type: resolveSignature(sig),
    });

    private visitParameter = (param: ParameterReflection): ITsMethodParameter => ({
        ...this.makeDocEntry(param, Kind.Parameter),
        defaultValue: getDefaultValue(param),
        type: resolveTypeString(param.type),
    });

    /** Visits each child that passes the filter condition (based on options). */
    private visitChildren<T>(
        children: Reflection[],
        visitor: (def: Reflection) => T,
        comparator?: (a: T, b: T) => number,
    ): T[] {
        const { excludeNames = [], includeNonExportedMembers = false } = this.options;
        return children
            .filter(ref => (ref.flags.isExported || includeNonExportedMembers) && isNotExcluded(excludeNames, ref.name))
            .map(visitor)
            .sort(comparator);
    }

    /**
     * Converts a typedoc comment object to a rendered `IBlock`.
     */
    private renderComment(comment: Comment) {
        if (!comment) {
            return undefined;
        }
        let documentation = "";
        if (comment.shortText) {
            documentation += comment.shortText;
        }
        if (comment.text) {
            documentation += "\n\n" + comment.text;
        }
        if (comment.tags) {
            documentation +=
                "\n\n" +
                comment.tags
                    .filter(tag => tag.tagName !== "default" && tag.tagName !== "deprecated")
                    .map(tag => `@${tag.tagName} ${tag.text}`)
                    .join("\n");
        }
        return this.compiler.renderBlock(documentation);
    }
}

function getCommentTag(comment: Comment, tagName: string) {
    if (comment == null || comment.tags == null) {
        return undefined;
    }
    return comment.tags.find(tag => tag.tagName === tagName);
}

function getDefaultValue(ref: DefaultValueContainer): string | undefined {
    if (ref.defaultValue != null) {
        return ref.defaultValue;
    }
    const defaultValue = getCommentTag(ref.comment, "default");
    if (defaultValue !== undefined) {
        return defaultValue.text.trim();
    }
    return undefined;
}

function getFileName({ sources = [] }: Reflection): string | undefined {
    const source = sources[0];
    const fileName = source && source.file && source.file.fullFileName;
    // filename relative to cwd, so it can be saved in a snapshot (machine-independent)
    return fileName && relative(process.cwd(), fileName);
}

function getFlags(ref: Reflection): ITsFlags | undefined {
    if (ref === undefined || ref.flags === undefined) {
        return undefined;
    }
    const isDeprecated = getIsDeprecated(ref);
    const { isExported, isExternal, isOptional, isPrivate, isProtected, isPublic, isRest, isStatic } = ref.flags;
    return { isDeprecated, isExported, isExternal, isOptional, isPrivate, isProtected, isPublic, isRest, isStatic };
}

function getIsDeprecated(ref: Reflection) {
    const deprecatedTag = getCommentTag(ref.comment, "deprecated");
    if (deprecatedTag === undefined) {
        return undefined;
    }
    const text = deprecatedTag.text.trim();
    return text === "" ? true : text;
}

/** Returns true if value does not match all patterns. */
function isNotExcluded(patterns: Array<string | RegExp>, value?: string) {
    return value === undefined || patterns.every(p => value.match(p) == null);
}

/** Sorts static members (`flags.isStatic`) before non-static members. */
function sortStaticFirst<T extends ITsDocBase>({ flags: aFlags = {} }: T, { flags: bFlags = {} }: T) {
    if (aFlags.isStatic && bFlags.isStatic) {
        return 0;
    } else if (aFlags.isStatic) {
        return -1;
    } else if (bFlags.isStatic) {
        return 1;
    }
    return 0;
}
