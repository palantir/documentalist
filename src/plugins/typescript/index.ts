/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { Application } from "typedoc";
import { ICompiler, IFile, IPlugin, ITypescriptPluginData } from "../../client";
import { Visitor } from "./visitor";

export interface ITypescriptPluginOptions {
    /** Exclude files by the given pattern when a path is provided as source. Supports standard minimatch patterns. */
    exclude?: string;

    /**
     * Whether `private` fields should be included in the data.
     * This is disabled by default as `private` fields typically do not need to be publicly documented.
     * @default false
     */
    includePrivates?: boolean;

    /**
     * Whether members not marked `export` should be included in the data.
     * This is disabled by default as non-exported members typically do not need to be publicly documented.
     * @default false
     */
    includeNonExported?: boolean;
}

export class TypescriptPlugin implements IPlugin<ITypescriptPluginData> {
    private app: TypedocApp;
    public constructor(private options: ITypescriptPluginOptions = {}) {
        const { exclude, includePrivates = false } = options;
        this.app = new TypedocApp({
            exclude,
            excludePrivate: !includePrivates,
            ignoreCompilerErrors: true,
            logger: "none",
        });
    }

    public compile(files: IFile[], compiler: ICompiler): ITypescriptPluginData {
        const project = this.getTypedocProject(files.map(f => f.path));
        const visitor = new Visitor(compiler, this.options);
        const typescript = compiler.objectify(visitor.visitProject(project), i => i.name);
        return { typescript };
    }

    private getTypedocProject(files: string[]) {
        const expanded = this.app.expandInputFiles(files);
        return this.app.convert(expanded);
    }
}

// tslint:disable-next-line:max-classes-per-file
class TypedocApp extends Application {
    // this tricks typedoc into working
    get isCLI() {
        return true;
    }
}
