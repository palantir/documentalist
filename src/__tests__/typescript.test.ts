/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { Documentalist } from "../documentalist";
import { TypescriptPlugin } from "../plugins/typescript/index";

describe("TypescriptPlugin", () => {
    const dm = Documentalist.create().use(".ts", new TypescriptPlugin());

    snapshot("classes");
    snapshot("interfaces");

    function snapshot(name: string) {
        it(`${name} snapshot`, async () => {
            const { typescript } = await dm.documentGlobs(`src/__tests__/__fixtures__/${name}.ts`);
            expect(typescript).toMatchSnapshot();
        });
    }
});
