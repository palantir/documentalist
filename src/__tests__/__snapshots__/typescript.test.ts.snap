// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypescriptPlugin classes snapshot 1`] = `
Object {
  "Animal": Object {
    "documentation": Object {
      "contents": Array [
        "<p>Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
Licensed under the BSD-3 License as modified (the “License”); you may obtain
a copy of the license in the LICENSE and PATENTS files in the root of this
repository.</p>
",
      ],
      "contentsRaw": "Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
Licensed under the BSD-3 License as modified (the “License”); you may obtain
a copy of the license in the LICENSE and PATENTS files in the root of this
repository.",
      "metadata": Object {},
    },
    "fileName": "src/__tests__/__fixtures__/classes.ts",
    "kind": "class",
    "methods": Array [
      Object {
        "fileName": "src/__tests__/__fixtures__/classes.ts",
        "kind": "method",
        "name": "bark",
        "signatures": Array [
          Object {
            "documentation": undefined,
            "kind": "signature",
            "parameters": Array [],
            "returnType": "string",
            "type": "() => string",
          },
        ],
      },
      Object {
        "fileName": "src/__tests__/__fixtures__/classes.ts",
        "kind": "method",
        "name": "eat",
        "signatures": Array [
          Object {
            "documentation": Object {
              "contents": Array [
                "<p>Public method.</p>
",
              ],
              "contentsRaw": "Public method.",
              "metadata": Object {},
            },
            "kind": "signature",
            "parameters": Array [
              Object {
                "defaultValue": undefined,
                "documentation": Object {
                  "contents": Array [
                    "<p>Name of the food to eat.</p>
",
                  ],
                  "contentsRaw": "Name of the food to eat.",
                  "metadata": Object {},
                },
                "fileName": "src/__tests__/__fixtures__/classes.ts",
                "flags": Array [],
                "kind": "parameter",
                "name": "food",
                "type": "string",
              },
            ],
            "returnType": "void",
            "type": "(food: string) => void",
          },
        ],
      },
    ],
    "name": "Animal",
    "properties": Array [],
  },
}
`;

exports[`TypescriptPlugin interfaces snapshot 1`] = `
Object {
  "IButtonProps": Object {
    "documentation": undefined,
    "fileName": "src/__tests__/__fixtures__/interfaces.ts",
    "kind": "interface",
    "methods": Array [],
    "name": "IButtonProps",
    "properties": Array [
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>If set to <code>true</code>, the button will display in an active state.
This is equivalent to setting <code>className=&quot;pt-active&quot;</code>.</p>
",
            Object {
              "tag": "undefined",
              "value": "false",
            },
          ],
          "contentsRaw": "If set to \`true\`, the button will display in an active state.
This is equivalent to setting \`className=\\"pt-active\\"\`.

@undefined false",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "active",
        "type": "undefined | true | false",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Whether this action is non-interactive.</p>
",
          ],
          "contentsRaw": "Whether this action is non-interactive.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "disabled",
        "type": "undefined | true | false",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>A ref handler that receives the native HTML element backing this component.</p>
",
          ],
          "contentsRaw": "A ref handler that receives the native HTML element backing this component.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "elementRef",
        "type": "undefined | (ref: HTMLElement) => any",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Name of the icon (the part after <code>pt-icon-</code>) to add to the button.</p>
",
          ],
          "contentsRaw": "Name of the icon (the part after \`pt-icon-\`) to add to the button.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "iconName",
        "type": "IconName",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Visual intent color to apply to element.</p>
",
          ],
          "contentsRaw": "Visual intent color to apply to element.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "intent",
        "type": "Intent",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>If set to <code>true</code>, the button will display a centered loading spinner instead of its contents.
The width of the button is not affected by the value of this prop.</p>
",
            Object {
              "tag": "undefined",
              "value": "false",
            },
          ],
          "contentsRaw": "If set to \`true\`, the button will display a centered loading spinner instead of its contents.
The width of the button is not affected by the value of this prop.

@undefined false",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "loading",
        "type": "undefined | true | false",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Click event handler.</p>
",
          ],
          "contentsRaw": "Click event handler.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "onClick",
        "type": "undefined | (event: MouseEvent) => void",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Name of the icon (the part after <code>pt-icon-</code>) to add to the button.</p>
",
          ],
          "contentsRaw": "Name of the icon (the part after \`pt-icon-\`) to add to the button.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "rightIconName",
        "type": "IconName",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>Action text.</p>
",
          ],
          "contentsRaw": "Action text.",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "text",
        "type": "string",
      },
      Object {
        "defaultValue": undefined,
        "documentation": Object {
          "contents": Array [
            "<p>HTML <code>type</code> attribute of button. Common values are <code>&quot;button&quot;</code> and <code>&quot;submit&quot;</code>.
Note that this prop has no effect on <code>AnchorButton</code>; it only affects <code>Button</code>.</p>
",
            Object {
              "tag": "undefined",
              "value": "\\"button\\"",
            },
          ],
          "contentsRaw": "HTML \`type\` attribute of button. Common values are \`\\"button\\"\` and \`\\"submit\\"\`.
Note that this prop has no effect on \`AnchorButton\`; it only affects \`Button\`.

@undefined \\"button\\"",
          "metadata": Object {},
        },
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "property",
        "name": "type",
        "type": "undefined | string",
      },
    ],
  },
  "ICompiler": Object {
    "documentation": Object {
      "contents": Array [
        "<p>Each plugin receives a <code>Compiler</code> instance to aid in the processing of source files.</p>
",
      ],
      "contentsRaw": "Each plugin receives a \`Compiler\` instance to aid in the processing of source files.",
      "metadata": Object {},
    },
    "fileName": "src/__tests__/__fixtures__/interfaces.ts",
    "kind": "interface",
    "methods": Array [
      Object {
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "method",
        "name": "objectify",
        "signatures": Array [
          Object {
            "documentation": Object {
              "contents": Array [
                "<p>Converts an array of entries into a map of key to entry, using given
callback to extract key from each item.</p>
",
              ],
              "contentsRaw": "Converts an array of entries into a map of key to entry, using given
callback to extract key from each item.",
              "metadata": Object {},
            },
            "kind": "signature",
            "parameters": Array [
              Object {
                "defaultValue": undefined,
                "documentation": undefined,
                "fileName": "src/__tests__/__fixtures__/interfaces.ts",
                "flags": Array [],
                "kind": "parameter",
                "name": "array",
                "type": "T[]",
              },
              Object {
                "defaultValue": undefined,
                "documentation": undefined,
                "fileName": "src/__tests__/__fixtures__/interfaces.ts",
                "flags": Array [],
                "kind": "parameter",
                "name": "getKey",
                "type": "(item: T) => string",
              },
            ],
            "returnType": "{ [key: string]: T }",
            "type": "(array: T[], getKey: (item: T) => string) => { [key: string]: T }",
          },
        ],
      },
      Object {
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "method",
        "name": "renderBlock",
        "signatures": Array [
          Object {
            "documentation": Object {
              "contents": Array [
                "<p>Render a block of content by extracting metadata (YAML front matter) and
splitting text content into markdown-rendered HTML strings and <code>{ tag,
value }</code> objects.</p>
<p>To prevent special strings like &quot;@include&quot; from being parsed, a reserved
tag words array may be provided, in which case the line will be left as
is.</p>
",
              ],
              "contentsRaw": "Render a block of content by extracting metadata (YAML front matter) and
splitting text content into markdown-rendered HTML strings and \`{ tag,
value }\` objects.

To prevent special strings like \\"@include\\" from being parsed, a reserved
tag words array may be provided, in which case the line will be left as
is.",
              "metadata": Object {},
            },
            "kind": "signature",
            "parameters": Array [
              Object {
                "defaultValue": undefined,
                "documentation": undefined,
                "fileName": "src/__tests__/__fixtures__/interfaces.ts",
                "flags": Array [],
                "kind": "parameter",
                "name": "blockContent",
                "type": "string",
              },
              Object {
                "defaultValue": undefined,
                "documentation": undefined,
                "fileName": "src/__tests__/__fixtures__/interfaces.ts",
                "flags": Array [
                  "Optional",
                ],
                "kind": "parameter",
                "name": "reservedTagWords",
                "type": "string[]",
              },
            ],
            "returnType": "object",
            "type": "(blockContent: string, reservedTagWords: string[]) => object",
          },
        ],
      },
      Object {
        "fileName": "src/__tests__/__fixtures__/interfaces.ts",
        "kind": "method",
        "name": "renderMarkdown",
        "signatures": Array [
          Object {
            "documentation": Object {
              "contents": Array [
                "<p>Render a string of markdown to HTML, using the options from <code>Documentalist</code>.</p>
",
              ],
              "contentsRaw": "Render a string of markdown to HTML, using the options from \`Documentalist\`.",
              "metadata": Object {},
            },
            "kind": "signature",
            "parameters": Array [
              Object {
                "defaultValue": undefined,
                "documentation": undefined,
                "fileName": "src/__tests__/__fixtures__/interfaces.ts",
                "flags": Array [],
                "kind": "parameter",
                "name": "markdown",
                "type": "string",
              },
            ],
            "returnType": "string",
            "type": "(markdown: string) => string",
          },
        ],
      },
    ],
    "name": "ICompiler",
    "properties": Array [],
  },
}
`;
